transport/http/client.go

```diff
@@ -84,7 +91,21 @@ func (c Client) Endpoint() endpoint.Endpoint {
 		ctx, cancel := context.WithCancel(ctx)
 		defer cancel()
 
-		req, err := http.NewRequest(c.method, c.tgt.String(), nil)
+		// Vars used for client finalizer to ensure there are no nil values
+		var (
+			req  *http.Request  = &http.Request{}
+			resp *http.Response = &http.Response{}
+			err  error
+		)
+		if c.finalizer != nil {
+			defer func() {
+				ctx = context.WithValue(ctx, ContextKeyResponseHeaders, resp.Header)
+				ctx = context.WithValue(ctx, ContextKeyResponseSize, resp.ContentLength)
+				c.finalizer(ctx, resp.StatusCode, req)
+			}()
+		}
```

Comment:
I'm not sure it's correct to create zero-value http.Requests and Responses; my intuition is that they need to be created by package http in an e.g. NewReqeust constructor. But that's fine, I think we can get by without them. Going with the change in the ClientFinalizerFunc below, perhaps this can look like..

```go
var (
    resp *http.Response
    err  error
)
if c.finalizer != nil {
    defer func() {
        if resp != nil {
            ctx = context.WithValue(ctx, ContextKeyResponseHeaders, resp.Header)
            ctx = context.WithValue(ctx, ContextKeyResponseSize, resp.ContentLength)
        }
        c.finalizer(ctx, err)
    }()
}
```

What do you think?